---
title: "HPA-gene-clustering"
author: "MarIa Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r setup, include=FALSE}

# Packages
library(tidyverse)
library(NOISeq)
library(uwot)
library(pcaMethods)
library(magrittr)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(fclust)
library(flashClust)
library(pheatmap)
library(clValid)
library(ggthemes)
library(ggrepel)
library(readr)
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
#source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select

```

```{r setup, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
  

# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```


#1- Data pre-processing

```{r pre-procesing}

# Filtering genes
genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","min-max","max") %>%
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

#scaled_data <- scaled [[1]]

a <- filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id")

gene_names <- a$enssscg_id

# Remove data
rm(a)
rm(data)
rm(tmm_data)
```

#2- Dimensionality reduction

```{r dimensionality reduction}

#All PCAs
pcas <- scaled %>%
  map(~pca_calc(data = .x, npcs = 200))

#pcas %>%
 # map(~pca_plot(data = .x))

# One pca
#pca_data <- 
 # scaled_data %>%
  #pca_calc(npcs = 200)

#pca_data %>%
 # pca_plot()

# Euc, Man, Pear

#scaled_data %>%
 # umap_calc(row_names = "enssscg_id", n_neigh = 20, color_tissue = T)
```

#3- Distance calculation

```{r pre-procesing}

#All distances
pcas, list("euclidean","manhattan","pearson")

combinations <- data.frame(pca = lits(pcas[[1]],pcas[[1]],pcas[[1]],
                     pcas[[2]],pcas[[2]],pcas[[2]],
                     pcas[[3]],pcas[[3]],pcas[[3]]),
             dist_method = c("euclidean","manhattan","pearson","euclidean","manhattan",
                             "pearson","euclidean","manhattan","pearson"))

distances <- 
  map2(combinations$pca, combinations$dist_method,
      ~dist_calc(df = .x$scores, comp = 30, m = .y))


d <- lapply(list("euclidean","manhattan","pearson"), function(i) map(pcas, 
                                                                ~dist_calc(df = .x$scores, 
                                                                           comp = 30, 
                                                                           m = i)))
            
            
            
save(d,
     file = "data/processed/all_distances.Rdata")


#One distance matrix

dist_to_cluster <-
  pca_data$scores %>%
  dist_calc(comp = 30, m = "euclidean")

save(dist_to_cluster,
     file = "data/processed/distance_zscore.Rdata")
#dist_to_cluster %>%
 # fviz_dist()

#dist_to_cluster %>%
 # pheatmap()
```

#4- Clustering

```{r cluster analysis}
res_kmeans_1 <- 
  dist_to_cluster %>%
  clust(k = 100, m = "kmeans")

results_kmeans <- results_kmeans
class(d[[1]])


results_kmeans <- lapply(list(1,2,3), function(i) map(d[[i]], 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fastkmeans"))
            
            
                                           df = .x$scores, 
                                                                           comp = 30, 
                                                                           m = i))


save(res_kmeans_1,
     file = "data/processed/clustering_zscore.Rdata")

res_clara %>%
  fviz_cluster()


load("data/processed/distance_zscore.Rdata")

res_fast <- 
  dist_to_cluster %>%
  clust(k = 100, m = "fasthclust")

res_kmeans_1$merge
res_kmeans_1 %>%
  cutree(100) %>%
  pheatmap()

fviz_cluster()
plot(res_kmeans_1)

fviz_dend(res_kmeans_1, cex = 0.5, k = 100,
 color_labels_by_k = FALSE, rect = TRUE)


res <- flashClust(dist_to_cluster, method = "ward", members = NULL)
res <- res %>% cutree(100) %>% as_tibble()

library(tidyverse)
library(broom)
library(magrittr)


  as_tibble(rownames = "gene") 

%>%
  
library(tidyverse)

%>%
      column_to_rownames("gene")

res$value = as.factor(res$value)
    
    pdf("dendogram.pdf")
    dist %>%
      pheatmap(annotation_row = res)

    res %>%
      fviz_dend(cex = 0.5, k = k)
    dev.off()


```

```{r}

load("data/processed/distance_zscore.Rdata")

res_kmeansfast <- 
  dist_to_cluster %>%
  clust(k = 100, m = "fastkmeans", genes = gene_names)



clus_umap <- scaled_data %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)


clus_umap$umap %>%
  left_join(res_kmeansfast, by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2, color = as.factor(cluster))) +
  geom_point() +
  theme_minimal() +
  coord_fixed() 

res_kmeansfast %>%
  fviz_cluster()

scaled_data

class(res_kmeansfast) <- "numeric"
```

#5- Evaluation

```{r pre-procesing}

# Orthologs

ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id)

X %>%
  left_join(ortholog_info)

# set orthologs to rownames

# evaluate


```

#6- Visualization
#7- Annotation