---
title: "HPA-gene-clustering"
author: "Maria Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r load packages and functions, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(NOISeq)
library(uwot)
library(MASS)
library(viridis)
library(pcaMethods)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(fclust)
library(flashClust)
library(dbscan)
library(igraph)
library(kmed)
library(aricode)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(ggthemes)
library(ggrepel)
library(readr)
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select

```

```{r load data, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```

#1- Data pre-processing

```{r pre-procesing}

# Filtering genes
filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Save gene names in order
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

#save(gene_names,
 #    file = "data/processed/gene_names.Rdata")

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","max") %>% #"min-max",
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

```

#2- Dimensionality reduction

```{r dimensionality reduction}

#All PCAs
#load("data/processed/scaled_data.Rdata")

pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

#save(pcas,
 #    file = "data/processed/pcas.Rdata")

```

#3- Distance calculation

```{r distance calculation}

#All distances
distances <- lapply(list("euclidean","pearson"), function(i) #"manhattan",
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))
            
#save(d,
 #    file = "data/processed/all_distances.Rdata")
```

#4- Clustering

```{r Clustering analysis}
load("data/processed/distances.Rdata")
load("data/processed/gene_names.Rdata")

d <- list(distances[[1]][[1]],distances[[1]][[2]],distances[[2]][[1]],distances[[2]][[2]])

res_kmeansfast <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "fastkmeans", genes = gene_names, id = .x$id))


res_kmeansfast <- 
  
kmeans_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
  map(d, ~clust(dist = .x$distance, k = as.numeric(i), 
                        m = "fastkmeans", genes = gene_names, id = .x$id, mult_k = T)))

kmeans_fasthclust <- 
  lapply(list(50,70,90,110,130,150), function(i)
  map(d, ~clust(dist = .x$distance, k = as.numeric(i), 
                        m = "fasthclust", genes = gene_names, id = .x$id, mult_k = T)))






kmeans_multk[[6]][[4]]$id

save(kmeans_multk,
     file = "data/processed/results_kmeansfast_multk_2.Rdata")

res_hclust <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "fasthclust", genes = gene_names,  id = .x$id))

save(res_hclust,
     file = "data/processed/results_hclust.Rdata")

res_medoids <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "medoids", genes = gene_names, id = .x$id))

save(res_medoids,
     file = "data/processed/results_medoids.Rdata")


clust_louvain <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "louvain",
                                                genes  = gene_names,
                                                id = .x$id))


clust_leiden <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "leiden",
                                                genes  = gene_names,
                                                id = .x$id))

clust_dbscan <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "dbscan",
                                                genes  = gene_names,
                                                id = .x$id))

clust_SOM <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "SOM",
                                                genes  = gene_names,
                                                id = .x$id))

```

```{r fuzzy tests}

library(advclust)


fuzzy <- fuzzy.CM(X = distances[[1]]$distance %>% as.matrix(), 
                  K = 100, m = 2, max.iteration = 100, threshold = 1e-5)


centroids <- KMeans_arma(dist %>% as.matrix(), clusters = k,
                             n_iter = 10, seed_mode = "random_subset")

res <- predict_KMeans(dist %>% as.matrix(), centroids)
    class(res) <- "numeric"
    
    
r <- KMeans_rcpp(
  distances[[1]]$distance %>% as.matrix(),
  clusters = 100,
  num_init = 1,
  max_iters = 100,
  initializer = "kmeans++",
  fuzzy = TRUE)


library(e1071)
c <-  cmeans(x = d$distance %>% as.matrix(), centers = 50,
        method="cmeans")


d <- distances[[1]]
b <- d %>% as.matrix() %>% as.data.frame()
library(RcmdrPlugin.FuzzyClust)

fuzzzz <- RcmdrPlugin.FuzzyClust::fuzzy.CM(X =  d$distance %>% as.matrix() %>% as.data.frame(), K = 100, m = 2, max.iteration = 100, threshold = 10^-5,
  RandomNumber = 0)
Sys.time()

library(advclust)
fuzzy <- advclust::fuzzy.CM(X = d$distance %>% as.matrix(), 
                  K = 100, m = 2, max.iteration = 100, threshold = 1e-5)
Sys.time() # Iterations .... No

res_fuz <- KMeans_rcpp(data = distances[[1]]$distance %>% as.matrix(), 
                       clusters = 100, fuzzy = TRUE)


res_fuzzyc1 <- 
  distances[[1]]$distance %>%
  clust(k = 100, m = "fuzzyc", genes = gene_names)
      


clust_fuzzyc <-  map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fuzzyc",
                                                genes  = gene_names,
                                                id = .x$id))

library(ppclust)
Sys.time()
x <- mfpcm(x = d$distance %>% as.matrix(), centers = 100)
Sys.time() # long time


library(Mfuzz)
?mfuzz
fuzz <- mfuzz(x = distances[[1]]$distance)
Sys.time()
f <- fanny(x = distances[[1]]$distance, k = 100, diss = TRUE,
      stand = FALSE, memb.exp = 1)
Sys.time()


library(e1071)
a <- cshell(x = distances[[1]]$distance %>% as.matrix(), centers = 100, iter.max=10, verbose=FALSE, dist="euclidean",
       method="cshell", m=2, radius = NULL)


```

```{r dbscan / SOM tests}
  
res_dbscan <- distances[[1]]$distance %>% dbscan::dbscan(eps = 100, minPts = 5)

res_som <- som(distances[[1]]$distance %>% as.matrix(), )

# DBSCAN on 30 components
d <- distances[[1]]
res_dbscan <- d$distance %>% dbscan::dbscan(eps = .1, minPts = 5)

# DBSCAN on 2 UMAP components
clus_umap <- pcas[[1]]$scores %>%
  as.data.frame() %>%
  select(1:30) %>%
  set_rownames(gene_names) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)

dist <- dist_calc(df = clus_umap$umap,
                  comp = 2,
                  m = "euclidean", 
                  id = "zscore")

res_dbscan_2 <- dist$distance %>% dbscan::dbscan(eps = 0.02, minPts = 10)

res1 <- as.numeric(res_dbscan$cluster)
res2 <- as.numeric(res_dbscan_2$cluster)

df<- data.frame(gene = gene_names, c1 = res1, c2 = res2)

umaps[[1]]$umap %>%
  left_join(df, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(c2))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() #+
#  scale_color_manual(values = pal) 

load("data/processed/pcas.Rdata")
load("data/processed/umaps.Rdata")

?som

res_som2 <- som(dist$distance %>% as.matrix(), grid = somgrid(10, 10, "hexagonal")) #9
res <- res_som$unit.classif
```

#5- Evaluation

```{r load data}
load("data/processed/results_kmeansfast.Rdata")
load("data/processed/results_medoids.Rdata")
load("data/processed/results_hclust.Rdata")
load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")
load("data/processed/results_som.Rdata")
load("data/processed/distances.Rdata")
load("data/processed/gene_names.Rdata")
load("data/processed/data_scaled.Rdata")

# Ortholog information
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id) %>%
  dplyr::mutate(entrez = mapIds(org.Hs.eg.db,
                  keys= ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL"))

```

```{r generate lists of clusters and evaluate} 
# Attach random clustering to cluster list
random <- random_cluster(100, 22007, names = gene_names)


id <- "random"
id <- "zscore euclidean"
id <- "zscore pearson"
id <- "max euclidean"
id <- "max pearson"







# Create dataframe with clustering results
l <- c(res_hclust,res_kmeansfast,clust_medoids, clust_leiden,clust_louvain, clust_SOM, list(random))

eval_res <- map(l, ~eval(clustering = .x, 
                     genes= ortholog_info,
                     dist = distances
                     ))
df2 <- data.frame()
for (eval in eval_res) {
  df2 <- bind_rows(df2,eval)
}

df2 <- df2 %>% mutate(time = as.numeric(time, units = "secs"))
ev_res <- df2


l_hclust <- c()
for (i in c(1:6)){
  l_hclust <- c(l_hclust,kmeans_fasthclust[[i]])
}

l_hclust[[1]]$id

load("data/processed/results_kmeansfast_multk_2.Rdata")


eval_hclust <- map(l_hclust, ~eval(clustering = .x, 
                     genes= ortholog_info,
                     dis = d
                     ))

df_eval_hclust <- data.frame()
for (eval in eval_hclust) {
  df_eval_hclust <- bind_rows(df_eval_hclust,eval)
}
df_eval_hclust <- df_eval_hclust %>% mutate(time = as.numeric(time, units = "secs"))


#res <-  ### clustering id double, rep (wrong time)
 # df$cluster_results %>%
  #group_by(id) %>%
  #do({
   # data <- .
    eval_res <- eval(clustering = data, 
                     genes= ortholog_info,
                     dist = distances,
                     times = df$cluster_time
                     #id = .$id,
                     )
  #})

eval_random <- eval(clustering= random %>% mutate(id="random"), 
                    genes = ortholog_info,
                    dist=distances)

load("data/processed/eval_random.Rdata")


# Attach random results
ev <- bind_rows(results,eval_random %>% dplyr::rename(id = clustering_id))

save(ev_res,
     file = "data/processed/ev_res.Rdata")

```

```{r test BHI}
r <- 
    random %>%
    select(gene,cluster) %>%
    left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
    na.omit() %>%
    dplyr::select(-gene) 

clusters <- as.numeric(as.character(r$cluster))
length(clusters)

test <- rep(c(1),times=147)
test2 <- c(1:147)

x <- org.Hs.egGO
library(org.Hs.egGO)
if(require("Biobase") && require("annotate") && require("GO.db") &&
       require("org.Hs.eg.db")) {
      BHI(as.numeric(as.character(test)), annotation= "org.Hs.eg.db", 
          names=as.character(as.numeric(r$entrez)), category="all")
}

r

goTerms <- annotate::getGO(as.character(as.numeric(r$entrez)), "org.Hs.eg.db")
  goTerms <- annotate::getGO(names, annotation)

bhi <- tapply(goTerms, as.numeric(as.character(r$cluster)),
              function(x) matchGO(x, "all"))

v <- c()
e <- tapply(goTerms, as.numeric(as.character(r$cluster)),
              function(x) c(v,x))

org.Hs.egGO

data(mouse)
express <- mouse[1:25,c("M1","M2","M3","NC1","NC2","NC3")]
rownames(express) <- mouse$ID[1:25]
## hierarchical clustering
Dist <- dist(express,method="euclidean")
clusterObj <- hclust(Dist, method="average")
nc <- 6 ## number of clusters
cluster <- cutree(clusterObj,nc)


library(moe430a.db)

if(require("Biobase") && require("annotate") && require("GO.db") &&
require("moe430a.db")) {
BHI(as.numeric(as.character(cluster)), 
    annotation="moe430a.db", names=rownames(express), category="all")
}

```

```{r pairwise ARI}
# Pairwise ARI - cluster comparison 

# Prepare data 
df2 <- res %>% select(-time,-clustering_id) %>% unique()
clusterings <- c(df2$id,"random")
to_eval <- df$cluster_results %>% bind_rows(random %>% mutate(id = "random"))
res_ari <- matrix(, nrow = 55, ncol = 55)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(to_eval %>% filter(id == i) %>% pull(cluster), 
             to_eval %>% filter(id == j) %>% pull(cluster))
    res_ari[i,j] <- a
    }
  }

# Pheatmap for all scalings, distances and clustering methods
res_ari %>%
  pheatmap(fontsize = 5)

# Pheatmap for specific scalings

d <- c()
for (id in clusterings) {
  if (grepl("max",id) & !grepl("min",id)){d <- c(d,id)}
}

d <- c()
for (id in clusterings) {
  if (grepl("pearson",id)){d <- c(d,id)}
}

res_ari %>%
  as_tibble() %>%
  mutate(id = clusterings)%>%
  filter(id %in% d) %>%
  select(d) %>%
  as.matrix() %>%
  set_rownames(d)%>%
  pheatmap(fontsize = 10)

```

```{r test different scaling for evaluation pheatmap}
ev %>%
  column_to_rownames("id") %>%
  mutate(connectivity_index = scale(connectivity_index)) %>%
  pheatmap(cluster_cols = F)

ev_minmax <- 
  ev %>%
  left_join(df$cluster_time %>% bind_rows(data.frame(id = "random", time = 1))) %>%
  column_to_rownames("id") %>%
  mutate(min = min(connectivity_index),
         max = max(connectivity_index),
         connectivity = (connectivity_index - min)/(max-min)) %>%
  select(connectivity,dunn_index,BHI_index, time) %>%
  mutate(min = min(dunn_index),
         max = max(dunn_index),
         dunn = (dunn_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI_index, time) %>%
  mutate(min = min(BHI_index),
         max = max(BHI_index),
         BHI = (BHI_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) %>%
   mutate(min = min(time),
         max = max(time),
         time = (time - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) 
  #mutate(sum_connectivity = scale(connectivity_index)) %>%

ev_minmax %>% 
  pheatmap(cluster_cols = F)


### UMAP clusterings

ev_umap <- 
  ev_minmax %>%
  select(-time) %>%
  rownames_to_column("id") %>%
  umap_calc(row_names = "id", n_neigh = 15, n_comp = 2)


scaling <- c()
distance <- c()
clustering <- c()

for (row in 1:nrow(ev_umap)) {
  elements <- strsplit(ev_umap[[row, "features"]], " ")
  scaling <- c(scaling, elements[[1]][1])
  distance <- c(distance, elements[[1]][2])
  clustering <- c(clustering, elements[[1]][3])
}

ev_umap_colors <-
  ev_umap %>%
  mutate(scaling = scaling,
         distance = distance,
         clustering = clustering)


ev_umap %>% 
  ggplot(aes(V1,V2, color =scaling)) +
      geom_point() +
      theme_minimal()+
      coord_fixed()
#color = viridis(10)
```

```{r plot evaluation results}

# Plots
r %>%
  ggplot(aes(x=id, y=dunn)) +
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) #+
#  scale_color_manual(values = pal) 


ev %>%
  column_to_rownames("id") %>%
  pheatmap() 

```

```{r rank evalutaion}

ev$rank_connectivity <-rank(-(ev$connectivity_index))
ev$rank_dunn <-rank((ev$dunn_index))
ev$rank_BHI <-rank(-(ev$BHI_index))

rank_eval <- ev %>% left_join(df$cluster_time %>% drop_na())
rank_eval$rank_time <-rank(rank_eval$time)

rank_eval %>%
  select(id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

ev2 <- ev %>%
  select(id, connectivity_index, dunn_index, )

scaling <- c()
distance <- c()
clustering <- c()

for (row in 1:nrow(rank_eval)) {
  elements <- strsplit(rank_eval[[row, "id"]], " ")
  scaling <- c(scaling, elements[[1]][1])
  distance <- c(distance, elements[[1]][2])
  clustering <- c(clustering, elements[[1]][3])
}

rank_eval$scaling <- scaling
rank_eval$distance <- distance
rank_eval$clustering <- clustering

rank_eval$scale_dist <- paste(scaling,distance)

rank_eval %>%
  as_tibble() %>%
  select(-id) %>%
  select(clustering, scale_dist, rank_dunn) %>%
  spread(key = scale_dist, value = rank_dunn) %>% 
  drop_na(clustering) %>%
  select(-`random NA`)%>%
  as.data.frame() %>%
  column_to_rownames("clustering") %>%
  pheatmap()


load("data/processed/ev_multk.Rdata")
load("data/processed/ev_res.Rdata")

ev_res$rank_connectivity <-rank((ev_res$connectivity_index))
ev_res$rank_dunn <-rank(-(ev_res$dunn_index))
ev_res$rank_BHI <-rank(-(ev_res$BHI_index))
ev_res$rank_time <-rank(ev_res$time)

ev_res %>%
  select(clustering_id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("clustering_id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))




df_eval_hclust$rank_connectivity <-rank((df_eval_hclust$connectivity_index))
df_eval_hclust$rank_dunn <-rank(-(df_eval_hclust$dunn_index))
df_eval_hclust$rank_BHI <-rank(-(df_eval_hclust$BHI_index))
df_eval_hclust$rank_time <-rank(df_eval_hclust$time)




df_eval_hclust$k <- c(50,50,50,50,70,70,70,70,90,90,90,90,110,110,110,110,130,130,130,130,150,150,150,150)
df_eval_hclust$id <- c(50,50,50,50,70,70,70,70,90,90,90,90,110,110,110,110,130,130,130,130,150,150,150,150)

for (i in (1:nrow(df_eval_hclust))) {
  df_eval_hclust$id[i] <- paste(df_eval_hclust$clustering_id[i],df_eval_hclust$k[i])
}

df_eval_hclust %>%
  select(id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

bestk <- 
  df_eval_hclust %>%
  mutate(total_rank = rank_connectivity + rank_dunn + rank_BHI) %>%
  mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
  group_by(id) %>%
  mutate(min = (min(total_rank))) %>%
  filter(total_rank == min)

all_best <- data.frame()

for (df in dfs) {
  b <- bestk %>%
    select(-rank_connectivity, -rank_dunn, -rank_BHI, -rank_time,-min,-total_rank, -k, -clustering_id)
  all_best <- bind_rows(all_best,b)
}
  
```

```{r}
to_plot <- df_eval_hclust %>%
  as_tibble() %>%
  mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
#  filter(id == "zscore euclidean fasthclust ") %>%
  select(id,k,connectivity_index,dunn_index,BHI_index,time) %>%
  gather(key = measure, value = value, connectivity_index:time) 

#%>%
  group_by(id) %>%
  do ({
    data <- .
    
    data %>%
      ggplot(aes(x=k, y=value)) +
      geom_point() +
      geom_line() +# coord_fixed() +
      theme_minimal() +
      facet_wrap(~measure, scales="free_y")
  })

plots <-
  lapply(unique(to_plot$id),
         function(i) {
           to_plot %>%
             filter(id == i)%>%
      ggplot(aes(x=k, y=value)) +
      geom_point() +
      geom_line() +# coord_fixed() +
      theme_minimal() +
      facet_wrap(~measure, scales="free_y")
         })


pdf("ev.pdf")
plots
dev.off()

```

#6- Visualization

```{r UMAP visualization}
load("data/processed/data_scaled.Rdata")
load("data/processed/pcas.Rdata")

clustering <- kmeans_multk[[1]][[1]]$cluster

clustering <- clust_louvain[[1]]$cluster
## UMAP

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)

clus_umap %>% umap_plot()
                        
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, fill = cluster, color = cluster)) +
  geom_point(size = 0.1) +
  stat_density2d(bins = 2, h = 2) +
  theme_bw()


geom_density_2d(bins = 2, )
  geom_density2d_filled(bins = 20)



#

## Hexagonal
  getmode <- function(v) {
  uniqv <- unique(v)
  as.character(uniqv[which.max(tabulate(match(v, uniqv)))])
}
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, z = cluster, group = 1)) +
  stat_summary_hex(fun = getmode,
                   bins = 100,
                   show.legend = F)
## Density
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, group=cluster)) +
  geom_density2d_filled(h = 10) +
  facet_wrap(~cluster)


#######

density_settings <-
  expand.grid(n = c(100),
              h = c(0.02, 0.05, 0.1, 0.2)) %>% 
  as_tibble()

clus_umap_density_all <- 
  density_settings %>% 
  group_by_all() %>% 
  do({
    h <- .$h
    n <- .$n
    
    clus_umap %>% 
      left_join(clustering %>% 
                  dplyr::rename(features = 1, 
                                cluster = 2)) %>% 
      mutate(cluster = factor(cluster)) %>%
      group_by(cluster) %>%
      do ({
        data <- .
        data %$%
          MASS::kde2d(V1, V2, n = n, h = h, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
          {.$z} %>%
          as_tibble(rownames = "x") %>%
          gather(y, density, -1) %>%
          mutate(y = gsub("V", "", y),
                 x = as.numeric(x),
                 y = as.numeric(y))
      })
  })
clus_umap_density <-
  clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  group_by(cluster) %>%
  do ({
    data <- .
    data %$%
      MASS::kde2d(V1, V2, n = 300, h = 0.1, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
      {.$z} %>%
      as_tibble(rownames = "x") %>%
      gather(y, density, -1) %>%
      mutate(y = gsub("V", "", y),
             x = as.numeric(x),
             y = as.numeric(y))
  })

clus_umap_density %>%
  
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) +
  
  geom_point(data = clus_umap,
             aes(V1, V2),
             size = 0.1,
             inherit.aes = F) + 
  geom_tile() +
  theme_bw() +
  scale_alpha_continuous(limits = c(0, 0.3), range = c(0.1, 0.6))


clus_umap_density %>%
  
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) 
  
clus_umap %>%
  ggplot() +
  geom_point(aes(V1, V2),
             size = 0.1,
             inherit.aes = F)

clus_umap_density %>%
  
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.005) %>%
  ggplot(aes(x,y,fill = cluster)) +
  geom_tile() +
  theme_bw() 
  
clus_umap %>%
  ggplot() +
  geom_point(aes(V1, V2),
             size = 0.1,
             inherit.aes = F)


clus_umap_density %>%
  filter(density > 0.1) %>% 
  group_by(x, y) %>%
  count %>% 
  group_by(n) %>% 
  count


# clus_umap_density %>%
#   filter(density>0.01) 
#   ggplot(aes(density)) +
#   geom_density() 



clus_umap %>% umap_plot(color_by = "cluster", 
                        color_groups =clust_leiden[[7]]$cluster)




### alluvial plot? compare clustering algorithms

to_all <- df %>% spread(key = id, value = cluster)  %>% as_tibble()


ids <- c("zscore euclidean fasthclust", "zscore euclidean louvain")
df_all <- df %>% filter(id %in% ids) UM %>% group_by(id,cluster)  
t<- df_all %>% spread(key = id, value = cluster)
e <- t %>% group_by(`zscore euclidean fasthclust`,`zscore euclidean louvain`, gene) %>% dplyr::count()

ggplot(data = e,
       aes(axis1 = `zscore euclidean fasthclust` , axis2 = `zscore euclidean louvain`,
           y = n)) +
  scale_x_discrete(limits = c("Cluster", "Panel"), expand = c(.2, .05)) +
  geom_alluvium(aes(fill = as.factor(`zscore euclidean fasthclust`)), show.legend = F) +
  geom_stratum() +
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme(axis.text.x = element_text(size = 18, face = "bold"),
            axis.text.y = element_blank(),
            axis.ticks = element_blank(),
            panel.background = element_blank(),
            axis.title = element_blank())  +
    coord_flip()

library(ggalt)
library(ggalluvial)
ggplot(data = df_all,
       aes(axis1 = id, axis2 = gene)) +
  scale_x_discrete(limits = 
                     c("max euclidean fastkmeans", "max euclidean fasthclust"), 
                   expand = c(.2, .05)) +
  geom_alluvium(aes()) + #fill= gene)
  geom_stratum() #+
  geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  theme(axis.text.x = element_text(size = 18, face = "bold"),
            axis.text.y = element_blank(),
            axis.ticks = element_blank(),
            panel.background = element_blank(),
            axis.title = element_blank()) 


```

```{r test differnt distance metrics in umap}
umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2)

umap_corr <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "correlation", color_tissue = T)

umap_manhattan <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "manhattan", color_tissue = T)

umap_euclidean <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "euclidean", color_tissue = T)

umaps <- list(umap_euclidean,umap_manhattan,umap_corr)

save(umaps,
     file = "data/processed/umaps.Rdata")

load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)

c <- clust(dist = d$distance, k = 100, m = "fastkmeans", genes = gene_names, 
           id = d$id)

umap$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)
c <- clust(dist = d$distance, k = 100, m = "louvain", genes = gene_names, 
           id = d$id)

umaps[[3]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 

```

```{Visual evaluation - test UMAP distance metrics (Louvain / Leiden)}

umap <- scaled[[1]]$scaled %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2)

umap$umap %>%
  left_join(res_kmeansfast[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() 

umap_corr <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "correlation")

umap_manhattan <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "manhattan")

umap_euclidean <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "euclidean")

umaps <- list(umap_euclidean,umap_manhattan,umap_corr)

save(umaps,
     file = "data/processed/umaps.Rdata")


load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")

# 2D clustering
d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)
c <- clust(dist = d$distance, k = 100, m = "fastkmeans", genes = gene_names, 
           id = d$id)

umap$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)

c <- clust(dist = d$distance, k = 100, m = "louvain", genes = gene_names, 
           id = d$id)

umaps[[3]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 



load("data/processed/umaps.Rdata")


df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)
umaps[[1]]$umap %>%
  left_join(df2, by = c("features" = "gene")) %>%
  ggplot(aes(V1, V2)) +
   geom_hex(aes(fill = 1), fill = "gray90") +
  theme_minimal() +
  geom_encircle(aes(fill = value,
                     color = value),
                 alpha = 0.1,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F) +
   geom_encircle(aes(color = value),
                 fill = NA,
                 alpha = 0.7,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F)# +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  
load("data/processed/results_louvain.Rdata")

df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)

umaps[[1]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  head(200) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal)  + 
    scale_fill_manual(values = pal)  + 

  #geom_density_2d()
#  stat_density_2d(aes(fill = value), geom="polygon")+ 
# stat_ellipse()
   geom_encircle(aes(group = value, fill = as.factor(value)),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                #fill = "gray",
               # color = "black",
                show.legend = F)  +
    geom_point(show.legend = F, size = 1) 
  #   ggplot(aes(V1, V2)) +
#   geom_hex(aes(fill = 1),
#            data = plot_data,
#            fill = "gray90",
#            bins = plot_bins) +
#   geom_encircle(aes(fill = merged_cluster,
#                     color = merged_cluster),
#                 alpha = 0.1,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_encircle(aes(color = merged_cluster),
#                 fill = NA,
#                 alpha = 0.7,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  

pal <- colorRampPalette(ocean.curl(50))(50)

  geom_encircle(aes(group = value),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                fill = "gray",
                color = "black",
                show.legend = F) 
#+ scale_color_manual(values = pal) 
```

``` {r test color palettes}
library(palettetown)
pal <- colorRampPalette(ichooseyou(pokemon = "Pichu"))(130) 

install.packages("palr")
library(palr)
pal <- colorRampPalette(sst_pal(130))(231)

library(pals)
pal <- colorRampPalette(ocean.curl(130))(130)

install.packages("jcolors")

library("jcolors")
pal <- colorRampPalette(jcolors::jcolors("pal8"))(130) 
pal <- colorRampPalette(tableau_color_pal("Tableau 20"))(100)

pal <- viridis_palette("Zissou1", 100, type = "continuous")
require(RColorBrewer)
brewer.pal(100, "Set2")

library(viridis)
pal <- viridis(100)

library(magma)
library(spe)
display.brewer.pal(100, "magma")
```


#7- Annotation

