---
title: "HPA-gene-clustering"
author: "María Bueno Álvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r setup, include=FALSE}

# Packages
library(tidyverse)
library(NOISeq)
library(uwot)
library(pcaMethods)
library(magrittr)
library(factoextra)
library(cluster)
library(kohonen)
library(pheatmap)
library(clValid)
library(ggthemes)
library(ggrepel)
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
#source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select

# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
  

# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```


#1- Data pre-processing

```{r pre-procesing}

# Filtering genes
genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","min-max","max") %>%
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))
scaled_data <- scaled [[1]]
```

#2- Dimensionality reduction

```{r dimensionality reduction}
pca_data <- 
  scaled_data %>%
  pca_calc(npcs = 200)

pca_data %>%
  pca_plot()

# Euc, Man, Pear

#scaled_data %>%
 # umap_calc(row_names = "enssscg_id", n_neigh = 20, color_tissue = T)
```

#3- Distance calculation

```{r pre-procesing}

dist_to_cluster <-
  pca_data$scores %>%
  dist_calc(comp = 30, m = "euclidean")

save(dist_to_cluster,
     file = "data/processed/distance_zscore.Rdata")
#dist_to_cluster %>%
 # fviz_dist()

#dist_to_cluster %>%
 # pheatmap()
```

#4- Clustering

```{r cluster analysis}
res_kmeans_1 <- 
  dist_to_cluster %>%
  clust(k = 100, m = "kmeans")

# 

save(res_kmeans_1,
     file = "data/processed/clustering_zscore.Rdata")

res_clara %>%
  fviz_cluster()



#memory.limit()
#memory.size()
```

#5- Evaluation

```{r pre-procesing}

# Orthologs

ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id)

X %>%
  left_join(ortholog_info)

# set orthologs to rownames

# evaluate


```

#6- Visualization

#7- Annotation