---
title: "HPA-gene-clustering"
author: "Maria Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r load packages and functions, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(readxl)
library(NOISeq)
library(uwot)
library(MASS)
library(viridis)
library(pcaMethods)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(fclust)
library(flashClust)
library(Seurat)
library(dbscan)
library(igraph)
library(kmed)
library(aricode)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(ggthemes)
library(ggrepel)
library(readr)
library(clusterProfiler)
library(mixOmics)

map <- purrr::map
pca <- pcaMethods::pca
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select
pca <- pcaMethods::pca

```

```{r load data, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```

#1- Data pre-processing

```{r pre-procesing}
filter <- dplyr::filter
# Filtering genes
filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)

# Save gene names in order
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

#save(gene_names,
 #    file = "data/processed/gene_names.Rdata")

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","max") %>% #"min-max",
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

```

#2- Dimensionality reduction

```{r dimensionality reduction}

#load("data/processed/scaled_data.Rdata")

pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

# save(pcas,
#      file = "data/processed/pcas.Rdata")

```

#3- Distance calculation

```{r distance calculation}

#All distances
distances <- lapply(list("euclidean","pearson"), function(i) #"manhattan"
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))

#save(distances,
 #    file = "data/processed/all_distances.Rdata")
```

#4- Clustering

```{r Clustering analysis}
#load("data/processed/distances_4.Rdata")
#load("data/processed/gene_names.Rdata")

#d <- list(distances[[1]][[1]],distances[[1]][[2]],distances[[2]][[1]],distances[[2]][[2]])

# Run with 1 k 
res_kmeans <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             k = 100, 
             m = "fastkmeans", 
             genes = gene_names, 
             id = .x$id))

# Run multi k
hclust_multk <- 
  lapply(list(50,70,90,110,130,150), function(i)
    map(d, ~clust(dist = .x$distance, 
                  k = as.numeric(i), 
                  m = "fasthclust", 
                  genes = gene_names, 
                  id = .x$id, 
                  mult_k = T)))

# Run with 1 r 
res_louvain <- 
  distances %>%
  map(~clust(dist = .x$distance, 
             r = 4, 
             m = "louvain", 
             genes = gene_names, 
             id = .x$id))

# Run multi r
res_louvain_multk <- 
  lapply(list(4,6,8,10,12,14), function(i)
    map(d, ~clust(dist = .x$distance, 
                  r = as.numeric(i), 
                  m = "louvain", 
                  genes = gene_names, 
                  id = .x$id)))

#save(hclust_multk,
 #    file = "data/processed/hclust_multk.Rdata")

```


#5- Evaluation

```{r load data}
load("data/processed/distances_4.Rdata")
load("data/processed/gene_names.Rdata")
load("data/processed/data_scaled.Rdata")

# Ortholog information
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id) %>%
  dplyr::mutate(entrez = mapIds(org.Hs.eg.db,
                  keys= ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL"))

```

```{r Create df for evaluation}
load("data/processed/distances_4.Rdata")
load("data/processed/leiden_multk.Rdata")

# Single evaluation
# test_eval <- eval(clustering = medoids_multk[[1]][[1]], 
#                   dis =  distances, 
#                   genes = ortholog_info)


eval_hclust_multk <- lapply(list(1,2,3,4,5,6), function(i) 
  map(hclust_multk[[i]], 
      ~eval(clustering = .x, 
            genes= ortholog_info,
            dis = d
      )))

#save(kmeans_multk, file = "data/processed/kmeans_multk.Rdata")
load("data/processed/eval_som_multk.Rdata")
load("data/processed/som_multk.Rdata")

eval_som_df <- eval_to_df(eval_somt_multk)


all_som <- add_clusterings(df = eval_som_df$res, 
                        dist = d, 
                        pca = pcas[[1]],
                        ref_clust = som_multk[[1]][[1]]$cluster,
                        gene_names = gene_names, 
                        clust_m = "SOM",
                        orthologs = ortholog_info)

bestk_leiden <- find_bestk(all_leiden)

save(eval_leiden_df,all_leiden,bestk_leiden, file = "data/processed/evaluation_leiden.Rdata")


#Plot all clustering
  # Create palette according to number of clusters

# Create the 4 alternative UMAPs


# Plots
clust_plot <- 
  function(clustering,umap)



```


```{r # Comparison to Max clustering}
# Max clustering
max_clustering <- 
  read_excel("data/processed/UMAP cluster annotation 20201126.xlsx", 
             sheet = "Cluster genes") %>%
  select(`Cluster nr`, `Ensembl 92 ID`) %>%
  rename(value = `Cluster nr`, 
         gene = `Ensembl 92 ID`) %>%
  mutate(value = as.numeric(value)) 

enriched_terms <- enrich(clustering = max_clustering, ontology_type = "BP", orthologs = ortholog_info)




test <- clust(dist = d[[1]]$distance, genes = gene_names, 
                   m = "fasthclust", k=50, id = d[[1]]$id)


#Random
clustering <- louvain_multk[[1]][[1]]$cluster
louvain_random <- clustering
louvain_random <-
  louvain_random %>%
  mutate(value = sample(value)) 

enriched_terms_random <- enrich(clustering = louvain_random, ontology_type = "BP", orthologs = ortholog_info)


# Louvain 66
clustering <- louvain_multk[[1]][[1]]$cluster 

enriched_terms_louvain <- enrich(clustering = clustering, ontology_type = "BP", orthologs = ortholog_info)
```

```{r # Comparison to Max clustering}
biological_indices <- data.frame(
  id=c("Random", "Louvain clustering", "2D kmeans clustering"), 
  biological_index = c(bio_index_random, bio_index_louvain, bio_index_max))
  #,
  p_values = c(list(pvalues_random), list(pvalues_max), list(pvalues_louvain)))


bio <- biological_indices %>%
  ggplot(aes(x=id, y=biological_index, group =1, fill = as.factor(id))) +
  geom_bar(stat = "identity", show.legend = F)+
  theme_minimal() +
  #ggtitle("Hierarchical clustering") +
  #facet_wrap(scale_dist~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=13)) + 
  labs(y = "BP enrichment", x = "Clustering strategy")


p_values <- bind_rows(data.frame(id = "Random", p_value = pvalues_random), 
                      data.frame(id = "2D kmeans clustering", 
                                           p_value = pvalues_max),
                      data.frame(id = "Louvain clustering", 
                                           p_value = pvalues_louvain))
pval<-
  p_values %>%
  #filter(p_value < 0.01) %>%
  ggplot(aes(x = id, y = -log10(p_value), fill = as.factor(id))) +
  geom_violin(show.legend = F, draw_quantiles = 0.5) +
  theme_minimal() +
  scale_fill_viridis(discrete = T) +
  scale_color_viridis(discrete = T)+
  theme(text = element_text(size=13)) + 
  labs(y = "p values", x = "Clustering strategy")

sort(pvalues_louvain)

bio + pval


intersect <- intersect(ck_max_tibble$Description, ck_louvain_tibble$Description)



  scale_fill_viridis(option="B", direction = -1) 

  p_values2 <- bind_rows(data.frame(clust_id = "Initial clustering", 
                                    p_value = ck_max_tibble$p.adjust,
                                    term = ck_max_tibble$Description),
                         data.frame(clust_id = "Alternative clustering", 
                                    p_value = ck_louvain_tibble$p.adjust,
                                    term = ck_louvain_tibble$Description)) %>%
  filter(term %in% intersect) %>%
  group_by(clust_id, term) %>%
  top_n(1, p_value)
  #select(clust_id, term, max(p_value))
    #3215

  #unique(p_values2$term)

  p_values2 %>%
    spread(key= clust_id, value= p_value) %>%
    ggplot(aes(x=-log10(`Alternative clustering`), y = -log10(`Initial clustering`))) +
    geom_point() +
    #scale_fill_viridis(discrete = T) +
    #scale_color_viridis(continuous = T)+
    theme(text = element_text(size=10)) + 
    #labs(y = "p values", x = "Clustering strategy") +
    geom_abline(slope = 1)+
    theme_minimal()

```

```{r Max UMAP}
load("data/processed/max_umap.Rdata")

max_clustering %>%
  left_join(ortholog_info) %>%
  select(value, enssscg_id) %>%
  left_join(pig_gene_umap_cl %>% as_tibble(rownames = "enssscg_id")) %>%
    #filter(value==i) %>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
   # geom_point(data = t, aes(V1,V2), inherit.aes = F, 
    #           size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5) +
    theme_minimal() +
    coord_fixed() +
  scale_color_manual(values = p2)

p2 <- sample(p2)
p2 <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(83)

louvain_random %>%
  left_join(ortholog_info) %>%
  select(value, enssscg_id) %>%
  left_join(pig_gene_umap_cl %>% as_tibble(rownames = "enssscg_id")) %>%
    #filter(value==i) %>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
   # geom_point(data = t, aes(V1,V2), inherit.aes = F, 
    #           size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5) +
    theme_bw() +
    coord_fixed() +
  scale_color_manual(values = p2) +
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))



```

```{r Plot metrics - multk}


# Barplot
df_louvain %>%
  as_tibble() %>%
  filter(scale_dist == 'zscore euclidean')%>%
  ggplot(aes(x=k, y=connectivity_index)) +
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


# Prepare data
to_plot <- df_hclust %>%
  as_tibble() %>%
  #mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
#  filter(id == "zscore euclidean fasthclust ") %>%
  select(scale_dist,k,connectivity_index,dunn_index, avg_silhouette_width, 
         BHI_index, time) %>% #BP_enriched
  gather(key = measure, value = value, connectivity_index:time) 

hclust_plot <- 
  to_plot %>%
  filter(measure != "time") %>%
  ggplot(aes(x=as.numeric(k), y=value, group =1, color = as.factor(measure))) +
  geom_point(show.legend = F) +
  geom_line(show.legend = F) +
  theme_minimal() +
  ggtitle("Hierarchical clustering") +
  facet_wrap(scale_dist~measure, scales="free_y"
             ) +
  scale_color_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  labs(y = "Value", x = "Number of clusters")

ggsave("ev_hclust.pdf", plot = hclust_plot)




# Rank
df_louvain <-
  df_louvain %>%
  mutate(rank_connectivity = rank(connectivity_index),
         rank_dunn = rank(-(dunn_index)),
         rank_asw = rank(-(avg_silhouette_width)),
         rank_BHI = rank(-(BHI_index)),
         rank_BPE = rank(-(BP_enriched)),
         rank_time = rank(time)
  )

# Select best k 
louvain_bestk <- 
  df_louvain %>%
  mutate(total_rank = rank_connectivity + rank_dunn + rank_asw + rank_BPE + rank_BHI) %>%
  #mutate(id = gsub('[[:digit:]]+', '', clustering_id)) %>%
  group_by(scale_dist) %>%
  mutate(min = (min(total_rank))) %>%
  filter(total_rank == min)

louvain_bestk_ranodm_2D <-
  bind_rows(
  louvain_bestk %>% select(clustering_id,connectivity_index, dunn_index,avg_silhouette_width, BHI_index, BP_enriched, time),
  ev_louvain_extra %>% select(clustering_id,connectivity_index, dunn_index,avg_silhouette_width, BHI_index, BP_enriched, time))


louvain_bestk_ranodm_2D %>%
  as_tibble() %>%
  gather(key = measure, value = value, connectivity_index:time) %>%
  select(-scale_dist) %>%
  #filter(scale_dist == 'zscore euclidean')%>%
  ggplot(aes(x=clustering_id, y=value, fill=clustering_id)) +
  geom_bar(stat = "identity")+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_wrap(~measure, scales="free_y") +
  scale_fill_viridis(discrete = T) +
  theme(text = element_text(size=10)) + 
  theme_minimal()
  labs(y = "Value", x = "Number of clusters")

ggsave("ev_hclust.pdf", plot = hclust_plot)

```

```{r Prepare all clusterings}

load("data/processed/results_kmeansfast.Rdata")
load("data/processed/results_medoids.Rdata")
load("data/processed/results_hclust.Rdata")
load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")
load("data/processed/results_som.Rdata")


```

```{r Plot metrics - different clusterings}
all_best <- data.frame()

for (df in dfs) {
  b <- bestk %>%
    select(-rank_connectivity, -rank_dunn, -rank_BHI, -rank_time,-min,-total_rank, -k, -clustering_id)
  all_best <- bind_rows(all_best,b)

  
# Generate ranking - plot heatmaps

ev_res %>%
  select(clustering_id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("clustering_id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))


df_eval_medoids$rank_connectivity <-
  rank(-(df_eval_medoids$connectivity_index))
df_eval_medoids$rank_dunn <-
  rank((df_eval_medoids$dunn_index))
df_eval_medoids$rank_BHI <-
  rank(-(df_eval_medoids$BHI_index))
df_eval_medoids$rank_time <-
  rank(df_eval_medoids$time)

df_eval_medoids %>%
  select(clustering_id,rank_connectivity, rank_dunn, rank_BHI, rank_time) %>%
  column_to_rownames("clustering_id") %>%
  pheatmap(cluster_cols = F, fontsize = 7, border_color = NA, 
           cutree_rows = 5, clustering_method = "ward.D2", 
           color = colorRampPalette(c("#00846b","white"))(100))

df_eval_medoids %>%
  as_tibble() %>%
  select(scale_dist,k,rank_connectivity) %>%
 # select(-clustering_id) %>%
  #select(clustering, scale_dist, rank_dunn) %>%
  spread(key = scale_dist, value = rank_connectivity) %>% 
  #d#rop_na(clustering) %>%
  #select(-`random NA`)%>%
  #select(k,`max euclidean`) %>%
  as.data.frame() %>%
  column_to_rownames("scale_dist") %>%
  pheatmap()
```

```{r test different scaling for evaluation pheatmap}
ev %>%
  column_to_rownames("id") %>%
  mutate(connectivity_index = scale(connectivity_index)) %>%
  pheatmap(cluster_cols = F)

ev_minmax <- 
  ev %>%
  left_join(df$cluster_time %>% bind_rows(data.frame(id = "random", time = 1))) %>%
  column_to_rownames("id") %>%
  mutate(min = min(connectivity_index),
         max = max(connectivity_index),
         connectivity = (connectivity_index - min)/(max-min)) %>%
  select(connectivity,dunn_index,BHI_index, time) %>%
  mutate(min = min(dunn_index),
         max = max(dunn_index),
         dunn = (dunn_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI_index, time) %>%
  mutate(min = min(BHI_index),
         max = max(BHI_index),
         BHI = (BHI_index - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) %>%
   mutate(min = min(time),
         max = max(time),
         time = (time - min)/(max-min)) %>%
  select(connectivity,dunn,BHI, time) 
  #mutate(sum_connectivity = scale(connectivity_index)) %>%

ev_minmax %>% 
  pheatmap(cluster_cols = F)


### UMAP clusterings

ev_umap <- 
  ev_minmax %>%
  select(-time) %>%
  rownames_to_column("id") %>%
  umap_calc(row_names = "id", n_neigh = 15, n_comp = 2)


scaling <- c()
distance <- c()
clustering <- c()

for (row in 1:nrow(ev_umap)) {
  elements <- strsplit(ev_umap[[row, "features"]], " ")
  scaling <- c(scaling, elements[[1]][1])
  distance <- c(distance, elements[[1]][2])
  clustering <- c(clustering, elements[[1]][3])
}

ev_umap_colors <-
  ev_umap %>%
  mutate(scaling = scaling,
         distance = distance,
         clustering = clustering)


ev_umap %>% 
  ggplot(aes(V1,V2, color =scaling)) +
      geom_point() +
      theme_minimal()+
      coord_fixed()
#color = viridis(10)
```

```{r pairwise ARI}

# Prepare data 
df2 <- res %>% select(-time,-clustering_id) %>% unique()
clusterings <- c(df2$id,"random")
to_eval <- df$cluster_results %>% bind_rows(random %>% mutate(id = "random"))
res_ari <- matrix(, nrow = 55, ncol = 55)
rownames(res_ari) <- c(clusterings)
colnames(res_ari) <- c(clusterings)

# Calculate ARI matrix
for (i in clusterings) {
  for (j in clusterings) {
    a <- ARI(to_eval %>% filter(id == i) %>% pull(cluster), 
             to_eval %>% filter(id == j) %>% pull(cluster))
    res_ari[i,j] <- a
    }
  }

# Pheatmap for all scalings, distances and clustering methods
res_ari %>%
  pheatmap(fontsize = 5)

# Pheatmap for specific scalings

d <- c()
for (id in clusterings) {
  if (grepl("max",id) & !grepl("min",id)){d <- c(d,id)}
}

d <- c()
for (id in clusterings) {
  if (grepl("pearson",id)){d <- c(d,id)}
}

res_ari %>%
  as_tibble() %>%
  mutate(id = clusterings)%>%
  filter(id %in% d) %>%
  select(d) %>%
  as.matrix() %>%
  set_rownames(d)%>%
  pheatmap(fontsize = 10)

```

#6- Visualization

```{r UMAP visualization - tests with Max}
load("data/processed/data_scaled.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/pcas.Rdata")
load("data/processed/louvain_multk.Rdata")
load("data/processed/medoids_multk.Rdata")

clustering <- louvain_multk[[1]][[1]]$cluster


## UMAP

clus_umap <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:10)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2) 

clus_umap %>% umap_plot()
                        
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, fill = cluster, color = cluster)) +
  geom_point(size = 0.1) +
  stat_density2d(bins = 2, h = 2) +
  theme_bw()


geom_density_2d(bins = 2, )
  geom_density2d_filled(bins = 20)


## Hexagonal
  getmode <- function(v) {
  uniqv <- unique(v)
  as.character(uniqv[which.max(tabulate(match(v, uniqv)))])
}
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, z = cluster, group = 1)) +
  stat_summary_hex(fun = getmode,
                   bins = 100,
                   show.legend = F)
## Density
clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  ggplot(aes(V1, V2, group=cluster)) +
  geom_density2d_filled(h = 10) +
  facet_wrap(~cluster)


#######

density_settings <-
  expand.grid(n = c(100,200,300,400),
              h = c(0.02, 0.05, 0.1, 0.2)) %>% 
  as_tibble()


clus_umap_density_all <- 
  density_settings %>% 
  group_by_all() %>% 
  do({
    h <- .$h
    n <- .$n
    
    clus_umap %>% 
      left_join(clustering %>% 
                  dplyr::rename(features = 1, 
                                cluster = 2)) %>% 
      mutate(cluster = factor(cluster)) %>%
      group_by(cluster) %>%
      do ({
        data <- .
        data %$%
          MASS::kde2d(V1, V2, n = n, h = h, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
          {.$z} %>%
          as_tibble(rownames = "x") %>%
          gather(y, density, -1) %>%
          mutate(y = gsub("V", "", y),
                 x = as.numeric(x),
                 y = as.numeric(y))
      })
  })
clus_umap_density <-
  clus_umap %>% 
  left_join(clustering %>% 
              dplyr::rename(features = 1, 
                            cluster = 2)) %>% 
  mutate(cluster = factor(cluster)) %>%
  group_by(cluster) %>%
  do ({
    data <- .
    data %$%
      MASS::kde2d(V1, V2, n = 300, h = 0.1, lims = c(range(clus_umap$V1), range(clus_umap$V2))) %>%
      {.$z} %>%
      as_tibble(rownames = "x") %>%
      gather(y, density, -1) %>%
      mutate(y = gsub("V", "", y),
             x = as.numeric(x),
             y = as.numeric(y))
  })

clus_umap_density%>%
 # filter(n== 400, h == 0.05) %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) +
 # geom_encircle() +
  geom_point(data = clus_umap %>% 
               left_join(clustering, by = c("features" = "gene")), #%>% filter(value ==2),
             aes(V1, V2),
             size = 0.07,
             inherit.aes = F) + 
  geom_tile() +
  theme_bw() +
  scale_alpha_continuous(limits = c(0, 0.3), range = c(0.05, 0.6))
  #facet_wrap(~cluster)#+
 # facet_grid(h ~ n)


clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.01) %>%
  ggplot(aes(x,y,fill = cluster, alpha = density)) 
  
clus_umap %>%
  ggplot() +
  geom_point(aes(V1, V2),
             size = 0.1,
             inherit.aes = F) 

clus_umap_density %>%
  mutate(x = scales::rescale(x, range(clus_umap$V1)),
         y = scales::rescale(y, range(clus_umap$V2))) %>%
  group_by(cluster) %>%
  mutate(density = density/max(density)) %>%
  ungroup () %>%
  filter(density > 0.005) %>%
  ggplot(aes(x,y,fill = cluster)) +
  geom_tile()+
  theme_bw() +
  scale_fill_manual(values = pal)

clus_umap %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2,color=as.factor(value))) +
  geom_point(size=0.6) +
  theme_bw() +
  scale_color_manual(values = pal) #+
  oose#facet_wrap(~value)

library(pals)
pal <- colorRampPalette(ocean.curl(60))(60)


clus_umap_density %>%
  filter(density > 0.1) %>% 
  group_by(x, y) %>%
  count %>% 
  group_by(n) %>% 
  count


# clus_umap_density %>%
#   filter(density>0.01) 
#   ggplot(aes(density)) +
#   geom_density() 

```

```{Visual evaluation - test UMAP distance metrics (Louvain / Leiden)}

umap <- scaled[[1]]$scaled %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2)

umap$umap %>%
  left_join(res_kmeansfast[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() 

umap_corr <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "correlation")

umap_manhattan <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "manhattan")

umap_euclidean <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) %>%
  rownames_to_column("enssscg_id") %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 20, n_comp = 2, met = "euclidean")

umaps <- list(umap_euclidean,umap_manhattan,umap_corr)

save(umaps,
     file = "data/processed/umaps.Rdata")


load("data/processed/results_leiden.Rdata")
load("data/processed/results_louvain.Rdata")

# 2D clustering
d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)
c <- clust(dist = d$distance, k = 100, m = "fastkmeans", genes = gene_names, 
           id = d$id)

umap$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 


d <- dist_calc(df = umap$umap, comp = 2, m = "euclidean", 
               id = distances[[1]]$id)

c <- clust(dist = d$distance, k = 100, m = "louvain", genes = gene_names, 
           id = d$id)

umaps[[3]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  geom_point(show.legend = F) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal) 



load("data/processed/umaps.Rdata")


df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)
umaps[[1]]$umap %>%
  left_join(df2, by = c("features" = "gene")) %>%
  ggplot(aes(V1, V2)) +
   geom_hex(aes(fill = 1), fill = "gray90") +
  theme_minimal() +
  geom_encircle(aes(fill = value,
                     color = value),
                 alpha = 0.1,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F) +
   geom_encircle(aes(color = value),
                 fill = NA,
                 alpha = 0.7,
                 s_shape=1,
                 expand=0,
                 # color = "white",
                 show.legend = F)# +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  
load("data/processed/results_louvain.Rdata")

df2 <- data.frame(value = as.numeric(medoids$cluster), gene = gene_names)

umaps[[1]]$umap %>%
  left_join(clust_louvain[[1]]$cluster, by = c("features" = "gene")) %>%
  head(200) %>%
  ggplot(aes(x = V1, y = V2, color = as.factor(value))) +
  theme_minimal() +
  coord_fixed() +
  scale_color_manual(values = pal)  + 
    scale_fill_manual(values = pal)  + 

  #geom_density_2d()
#  stat_density_2d(aes(fill = value), geom="polygon")+ 
# stat_ellipse()
   geom_encircle(aes(group = value, fill = as.factor(value)),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                #fill = "gray",
               # color = "black",
                show.legend = F)  +
    geom_point(show.legend = F, size = 1) 
  #   ggplot(aes(V1, V2)) +
#   geom_hex(aes(fill = 1),
#            data = plot_data,
#            fill = "gray90",
#            bins = plot_bins) +
#   geom_encircle(aes(fill = merged_cluster,
#                     color = merged_cluster),
#                 alpha = 0.1,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_encircle(aes(color = merged_cluster),
#                 fill = NA,
#                 alpha = 0.7,
#                 s_shape=1,
#                 expand=0,
#                 # color = "white",
#                 show.legend = F) +
#   geom_text(data = . %>%
#               select(merged_cluster, mean_V1, mean_V2, n) %>%
#               distinct(),
#             aes(mean_V1, mean_V2, label = merged_cluster),
#             show.legend = F)
  
  

pal <- colorRampPalette(ocean.curl(50))(50)

  geom_encircle(aes(group = value),
                alpha = 0.1, 
                s_shape=1, 
                expand=0,
                fill = "gray",
                color = "black",
                show.legend = F) 
#+ scale_color_manual(values = pal) 
```

```{r Cluster visualziationn - UMAP}

umap_louv %>% 
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  filter(enhanced_tissues == "retina") %>%
  ggplot(aes(V1,V2)) +
  geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
  geom_point(show.legend = F, color = tissue_colors[["retina"]]) +
  
  theme_minimal() +
  #scale_color_manual(values = tissue_colors) +
  coord_fixed()

# For all tissues
tissues <- unique(tissue_info$enhanced_tissues)
t
plots <- 
  lapply(unique(tissue_info$enhanced_tissues),
         function(tissue) {
           umap_louv %>%
             left_join(tissue_info, by= c("features" = "ensg_id")) %>%
             filter(enhanced_tissues == tissue) %>%
            ggplot(aes(V1,V2)) +
              geom_point(data = t, aes(V1,V2), color = "grey", inherit.aes = F) +
              geom_point(show.legend = F, color = tissue_colors[[tissue]]) +
              theme_minimal() +
              coord_fixed() +
             ggtitle(tissue)
         })

pdf("umaps_tissues.pdf")
plots
dev.off()


# Tissues for each cluster

to_plot <- 
  umap_louv %>%
  left_join(tissue_info, by= c("features" = "ensg_id")) %>%
  left_join(clustering, by = c("features" = "gene")) %>%
  mutate (value = value+1)

t <-
  to_plot %>%
  select(-value,-enhanced_tissues)

tissue_colors[["NA"]] <- "grey30"

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(67)
plots <- 
  lapply(sort(unique(to_plot$value)),
         function(cluster) {
           to_plot %>%
             filter(value == cluster)%>%
             ggplot(aes(V1,V2,color=enhanced_tissues)) +
             geom_point(data = t, aes(V1,V2), inherit.aes = F, 
                        size = 0.5, color= "gray93") +
             geom_point(show.legend = F, size = 0.5) +
             theme_minimal() +
             coord_fixed() +
             ggtitle(paste("cluster",cluster)) + 
             scale_color_manual(values = tissue_colors) 
         })

pdf("us.pdf")
plots
dev.off()


```


#7- Annotation

```{r Enrichment: test BP function}
# Select clustering result to annotate
#clustering <- XXX$cluster

# Prepare clustering data
r <- 
   clustering %>%
    left_join(ortholog_info, by = c("gene" = "enssscg_id")) %>%
    dplyr::select(-gene,-ensg_id)%>%
    na.omit()

clusters <- as.numeric(as.character(r$value))
cluster_list <- c()
num_clus <- tail(sort(r$value), n=1)
num_genes <- length(r$value)
for (i in c(0:num_clus)) {
  genes <- r %>%
    filter(value==i) %>%
    pull(entrez) %>%
    as.character()
  print(i)
  cluster_list[[as.character(i)]] <- genes
}
  
# Profile clusters (with universe)
ck_nouni <- compareCluster(geneClusters = cluster_list, 
                     fun = "enrichGO", 
                     OrgDb = org.Hs.eg.db, 
                     ont = "BP", 
                     #universe = as.character(r$entrez),
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05, 
                    qvalueCutoff = 0.05)

cknouni_tibble <- as_tibble(ck_nouni) 

length(unique(as.numeric(ck_tibble$Cluster)))/num_clus
#0.969697 (no universe) - 0.9393939 (universe)

# Distribution of clusters
clustering %>% 
  group_by(value) %>%
  count

# Most significant term for each cluster
ck_tibble %>%
  group_by(Cluster) %>%
  summarize(min(p.adjust)) 

# Explore specific clusters
clus21 <- 
  ck_tibble %>%
  filter(Cluster == 21)

```

```{r UMAP plot by cluster (PCA / PLS)}
clustering <- louvain_multk[[1]][[1]]$cluster

# PCA gene matrix to UMAP (Alternative 1)
gene_matrix <- pcas[[1]]$scores %>%
  set_rownames(gene_names) %>%
  as.data.frame() %>%
  select((1:30)) 

umap_louv <- gene_matrix %>%
  select((1:30)) %>%
  rownames_to_column("gene") %>%
  umap_calc(n_neigh = 15, row_names = "gene")

# PLS gene matrix to UMAP (Alternative 2)
gene_matrix <- scaled[[1]]$scaled %>% column_to_rownames("enssscg_id") 

pls_louv <- plsda(X= gene_matrix, Y = clustering$value, ncomp = 30)

umap_pls_louv <- pls_louv$variates$X%>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  umap_calc(n_neigh = 15, row_names = "gene")


## COMMON PART
# Data to plot, create "grey" UMAP background
to_plot <- 
  umap_louv %>%
  left_join(clustering, by = c("features" = "gene"))
t <- to_plot %>%
  select(-value)

p <- colorRampPalette(tableau_color_pal(palette = "Classic Cyclic")(13))(67)

#Rename clusters 0->1
to_plot <- to_plot %>%
  mutate(value = value+1)

# Highlight each cluster
plots <-
  lapply(sort(unique(to_plot$value)),
function(i) {
  to_plot%>%
    filter(value==i) %>%
    ggplot(aes(V1,V2)) +
    geom_point(data = t, aes(V1,V2), inherit.aes = F, 
               size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5, color= p[i]) +
    theme_minimal() +
    coord_fixed() +
    ggtitle(paste("cluster",i))
})

pdf("umaps.pdf")
plots
dev.off()


# Plot 1 cluster
to_plot%>%
    filter(value==57) %>%
    ggplot(aes(V1,V2)) +
    geom_point(data = t, aes(V1,V2), inherit.aes = F, 
               size = 0.5, color= "grey") +
    geom_point(show.legend = F, size =0.5, color= p[57]) +
    theme_minimal() +
    coord_fixed()

# Distribution of genes in clusters
distr <- to_plot %>%
  group_by(value) %>%
  count()

to_plot%>%
    ggplot(aes(V1,V2, color= as.factor(value))) +
    geom_point(show.legend = F, size =0.4) +
    theme_minimal() +
    coord_fixed() +
  scale_color_manual(values = p)

```

```{r Tissue expression per cluster (pheatmaps)}
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


#clustering <- louvain_multk[[1]][[1]]$cluster
#lustering <- max_clustering

data_tissue <-
  tmm_data %>%
  mutate(tissue = gsub('.{2}$', '', sample_ID)) %>%
  group_by(tissue,enssscg_id) %>%
  mutate(tissue_exp = mean(tmm)) %>%
  select(enssscg_id, tissue, tissue_exp) %>%
unique()

clustering <-
  clustering %>%
  mutate(value = value+1)

pdf("heatmaps_Max.pdf")
lapply(sort(unique(clustering$value)),
         function(i) {
           plot <- data_tissue %>%
             left_join(clustering, by = c("enssscg_id" = "gene")) %>%
            # mutate(value = value+1) %>%
             filter(value == i) %>%
             group_by(enssscg_id) %>%
             mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
             ungroup() %>%
             select(enssscg_id,tissue,scaled_value) %>%
             spread(key=tissue, value = scaled_value) %>%
             column_to_rownames("enssscg_id") %>% 
             pheatmap(clustering_method = "ward.D2", 
                      color = viridis(20, direction = -1, option = "B"),
                      show_rownames = F,border_color = NA, fontsize = 8)
         })
dev.off()
```

```{r Intra cluster correlations}

data <- data_tissue %>%
  group_by(enssscg_id) %>%
  mutate(scaled_value = tissue_exp / max(tissue_exp)) %>%
  ungroup() %>%
  select(enssscg_id,tissue,scaled_value) %>%
  spread(key=tissue, value = scaled_value)

genes <- clustering %>%
  mutate(value = value +1) %>%
  filter(value==1) %>%
  pull(gene)

cor(matrix["ENSSSCG00000000176",],matrix["ENSSSCG00000000176",])
  
cor(t(matrix))

matrix <- data %>%
  filter(enssscg_id %in% genes) %>%
  column_to_rownames("enssscg_id")



# Set up gene-sample matrix (log tmm)
tmm_data %>%
  mutate(log_tmm = log(tmm))%>%
  left_join(clustering, by = c("ensssgd_ID" = "gene")) %>%
  group_by(value) %>%
  do({
    data <- .
    cor_matrix <- cor(data)
    
  })
# Calculate correlation matrix for all clusters


# Filter correlation matix



#Visualization

correlations %>%
  ggplot(aes(x = cluster, y = correlation)) +
  geom_violin(draw_quantiles = 0.5)

```