---
title: "HPA-gene-clustering"
author: "MarIa Bueno Alvez"
date: "2/11/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

#Setup

```{r setup, include=FALSE}

# Packages
library(tidyverse)
library(magrittr)
library(NOISeq)
library(uwot)
library(pcaMethods)
library(factoextra)
library(cluster)
library(kohonen)
library(ClusterR)
library(fclust)
library(flashClust)
library(dbscan)
library(pheatmap)
library(clValid)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(annotate)
library(GO.db)
library(ggthemes)
library(ggrepel)
library(readr)
#concaveman, ggalt, sf

# Wd, scripts
setwd("/Users/maria.bueno/Desktop/HPA-gene-clustering")
source("scripts/functions_preprocessing.R")
source("scripts/functions_dimensionalityreduction.R")
source("scripts/functions_clustering.R")
source("scripts/functions_evaluation.R")

map <- purrr::map
select <- dplyr::select

```

```{r setup, include=FALSE}
# Data
data <- 
  read_delim("data/pig_filtered_sample_data.tab", delim = "\t")

gene_info <-
  read_delim("data/meta/pig_consensus_category_piggenome_92.tsv", delim = "\t")

tissue_info <-
  gene_info %>%
  filter(!ensg_id %in% genes_to_filter) %>%
  separate_rows("enhanced_tissues", sep = ',') %>%
  select(ensg_id,enhanced_tissues) %>%
  remove_missing()
  
  

# Color palette
colors <-
  read_delim("data/meta/pig_tissue_colors.tsv", delim = "\t")

tissue_colors <- colors  %>%
  select(organ_color, consensus_tissue_name) %>%
  distinct() %$%
  set_names(organ_color, consensus_tissue_name)

```


#1- Data pre-processing

```{r pre-procesing}

# Filtering genes
genes_to_filter <-
  gene_info %>%
  filter(distribution_category == "not detected") %>%
  pull(ensg_id) 

filtered_data <-
  data %>% 
  filter(!enssscg_id %in% genes_to_filter)


# Save gene names in order
gene_names <- 
  filtered_data %>%
  select(enssscg_id, sample_ID, tpm) %>%
  spread(key = sample_ID , value = tpm) %$%
  enssscg_id

save(gene_names,
     file = "data/processed/gene_names.Rdata")

# Normalize
tmm_data <- 
  filtered_data %>% 
  ptpm_normalize(sample_col = "sample_ID", tpm_col= "tpm", format = "wide") %>%
  tmm_normalize(doWeighting = FALSE) %>%
  as_tibble(rownames = "enssscg_id") %>%
  gather(key = "sample_ID", value = "tmm", - enssscg_id)


# Scale data
scaled <- list("zscore","min-max","max") %>%
  map(~data_scaling(df = tmm_data, col_value = "tmm",
                    col_gene = "enssscg_id", col_sample = "sample_ID",
                    log_transform = F, m = .x))

# Remove data
rm(a)
rm(data)
rm(tmm_data)
```

#2- Dimensionality reduction

```{r dimensionality reduction}

#All PCAs
pcas <- scaled %>%
  map(~pca_calc(data = .x$scaled, npcs = 200, m = .x$method))

pcas[[1]]$method

#pcas %>%
 # map(~pca_plot(data = .x))

# One pca
#pca_data <- 
 # scaled_data %>%
  #pca_calc(npcs = 200)

#pca_data %>%
 # pca_plot()

# Euc, Man, Pear

#scaled_data %>%
 # umap_calc(row_names = "enssscg_id", n_neigh = 20, color_tissue = T)
```

#3- Distance calculation

```{r pre-procesing}

#All distances
distances <- lapply(list("euclidean","manhattan","pearson"), function(i)
  map(pcas, ~dist_calc(df = .x$scores, comp = 30, m = i, 
                       id = .x$method)))
            
save(d,
     file = "data/processed/all_distances.Rdata")


#One distance matrix
dist_to_cluster <-
  pca_data$scores %>%
  dist_calc(comp = 30, m = "euclidean")

save(dist_to_cluster,
     file = "data/processed/distance_zscore.Rdata")
#dist_to_cluster %>%
 # fviz_dist()

#dist_to_cluster %>%
 # pheatmap()
```

#4- Clustering

```{r cluster analysis}
res_kmeans_1 <- 
  dist_to_cluster %>%
  clust(k = 100, m = "kmeans")

results_kmeans <- results_kmeans
class(d[[1]])


results_kmeans <- lapply(list(1,2,3), function(i) map(d[[i]], 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fastkmeans",
                                                genes  = gene_names,
                                                id = .x$id)))

save(results_kmeans,
     file = "data/processed/clusterings_kmeans.Rdata")

res_clara %>%
  fviz_cluster()


load("data/processed/distance_zscore.Rdata")

res_fast <- 
  dist_to_cluster %>%
  clust(k = 100, m = "fasthclust")

res_kmeans_1$merge
res_kmeans_1 %>%
  cutree(100) %>%
  pheatmap()

fviz_cluster()
plot(res_kmeans_1)

fviz_dend(res_kmeans_1, cex = 0.5, k = 100,
 color_labels_by_k = FALSE, rect = TRUE)


res <- flashClust(dist_to_cluster, method = "ward", members = NULL)
res <- res %>% cutree(100) %>% as_tibble()

library(tidyverse)
library(broom)
library(magrittr)


  as_tibble(rownames = "gene") 

%>%
  
library(tidyverse)

%>%
      column_to_rownames("gene")

res$value = as.factor(res$value)
    
    pdf("dendogram.pdf")
    dist %>%
      pheatmap(annotation_row = res)

    res %>%
      fviz_dend(cex = 0.5, k = k)
    dev.off()


```

```{r}

load("../distances.Rdata")

res_kmeansfast <- 
  dist_to_cluster %>%
  clust(k = 100, m = "fastkmeans", genes = gene_names)


res_hclust <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "fasthclust", genes = gene_names))

res_fuzzyc1 <- 
  distances[[1]]$distance %>%
  clust(k = 100, m = "fuzzyc", genes = gene_names)
      
      
      
  res_hclust <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "fasthclust", genes = gene_names)
      
res_hclust <- 
  distances %>%
  map(~clust(dist = .x$distance, k = 100, m = "fasthclust", genes = gene_names)


      

clus_umap <- scaled_data %>%
  umap_calc(row_names = "enssscg_id", n_neigh = 15, n_comp = 2)


clus_umap$umap %>%
  left_join(res_kmeansfast, by = c("features" = "gene")) %>%
  ggplot(aes(V1,V2, color = as.factor(cluster))) +
  geom_point() +
  theme_minimal() +
  coord_fixed() 

res_kmeansfast %>%
  fviz_cluster()

scaled_data

class(res_kmeansfast) <- "numeric"


clust_fastkmeans <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fastkmeans",
                                                genes  = gene_names,
                                                id = .x$id))

clust_fuzzyc <-  map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fuzzyc",
                                                genes  = gene_names,
                                                id = .x$id))

clust_fasthclust <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "fasthclust",
                                                genes  = gene_names,
                                                id = .x$id))

library(dbsca)

clust_louvain <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "louvain",
                                                genes  = gene_names,
                                                id = .x$id))


clust_dbscan <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "dbscan",
                                                genes  = gene_names,
                                                id = .x$id))

clust_SOM <- map(distances, 
                                         ~clust(dist = .x$distance,
                                                k = 100,
                                                m = "SOM",
                                                genes  = gene_names,
                                                id = .x$id))
```

#5- Evaluation

```{r pre-procesing}

# Attach random clustering to cluster list



# Ortholog information
ortholog_info <- 
  read_delim("data/meta/gene_orthologs.csv", delim = ",") %>%
  select(enssscg_id, ensg_id)

ortholog_info$entrez = mapIds(org.Hs.eg.db,
                  keys=ortholog_info$ensg_id, 
                  column="ENTREZID",
                  keytype="ENSEMBL")

# Evaluation
res_eval <- 
  res_hclust[[1]]$cluster %>%
  eval(genes = ortholog_info, distance = distances[[1]]$distance %>% as.matrix())


# Plots

  

```

#6- Visualization
#7- Annotation